sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -151,6 +151,238 @@ object ConstantPropagation extends Rule[LogicalPlan] with PredicateHelper {
+        case c EqualNullSafe d if planEqual(a, c) => EqualTo(c, d)
+        case c EqualNullSafe d if planEqual(a, d) => EqualTo(c, d)
+      }
+      case a LessThanOrEqual b => expression transformUp {
+        case c LessThan d if planEqual(b, d) && valueLessThan(c, a) =>
+          Literal.TrueLiteral
+        case c LessThan d if planEqual(b, c) && (planEqual(a, d) || valueLessThanOrEqual(d, a)) =>
+          Literal.FalseLiteral
+        case c LessThan d if planEqual(a, c) && valueLessThan(b, d) =>
+          Literal.TrueLiteral
+        case c LessThan d if planEqual(a, d) && (planEqual(b, c) || valueLessThanOrEqual(b, c)) =>
+          Literal.FalseLiteral
+
+        case c LessThanOrEqual d
+          if planEqual(b, d) && (planEqual(a, c) || valueLessThanOrEqual(c, a)) =>
+          Literal.TrueLiteral
+        case c LessThanOrEqual d if planEqual(b, c) && valueLessThan(d, a) =>
+          Literal.FalseLiteral
+        case c LessThanOrEqual d if planEqual(b, c) && (planEqual(a, d) || valueEqual(a, d)) =>
+          EqualTo(c, d)
+        case c LessThanOrEqual d
+          if planEqual(a, c) && (planEqual(b, d) || valueLessThanOrEqual(b, d)) =>
+          Literal.TrueLiteral
+        case c LessThanOrEqual d if planEqual(a, d) && valueLessThan(b, c) =>
+          Literal.FalseLiteral
+        case c LessThanOrEqual d if planEqual(a, d) && (planEqual(b, c) || valueEqual(b, c)) =>
+          EqualTo(c, d)
+
+        case c EqualTo d if planEqual(b, d) && valueLessThan(c, a) => Literal.FalseLiteral
+        case c EqualTo d if planEqual(b, c) && valueLessThan(d, a) => Literal.FalseLiteral
+        case c EqualTo d if planEqual(a, c) && valueLessThan(b, d) => Literal.FalseLiteral
+        case c EqualTo d if planEqual(a, d) && valueLessThan(b, c) => Literal.FalseLiteral
+
+        case c EqualNullSafe d if planEqual(b, d) && valueLessThan(c, a) => Literal.FalseLiteral
+        case c EqualNullSafe d if planEqual(b, c) && valueLessThan(d, a) => Literal.FalseLiteral
+        case c EqualNullSafe d if planEqual(a, c) && valueLessThan(b, d) => Literal.FalseLiteral
+        case c EqualNullSafe d if planEqual(a, d) && valueLessThan(b, c) => Literal.FalseLiteral
+
+        case c EqualNullSafe d if planEqual(b, d) => EqualTo(c, d)
+        case c EqualNullSafe d if planEqual(b, c) => EqualTo(c, d)
+        case c EqualNullSafe d if planEqual(a, c) => EqualTo(c, d)
+        case c EqualNullSafe d if planEqual(a, d) => EqualTo(c, d)
+      }
+      case a EqualTo b => expression transformUp {
+        case c LessThan d if planEqual(b, d) && planEqual(a, c) => Literal.FalseLiteral
+        case c LessThan d if planEqual(b, c) && planEqual(a, d) => Literal.FalseLiteral
+        case c LessThan d if planEqual(a, d) && planEqual(b, c) => Literal.FalseLiteral
+        case c LessThan d if planEqual(a, c) && planEqual(b, d) => Literal.FalseLiteral
+
+        case c LessThanOrEqual d if planEqual(b, d) && planEqual(a, c) => Literal.TrueLiteral
```

Comment:
these are wrong if the expression return `null`

Reply:
fixed, please see details below